File: 3Sum.cpp
class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
          vector<vector< int > > result; 
          if(num.size()<3) return result; 
          sort(num.begin(), num.end());
          for(int i=0; i<=num.size()-3; i++)
          {
              int first=num[i];
              if(i!=0 && num[i]==num[i-1]) continue; 
              int j=i+1, k=num.size()-1; 
              while(j<k)
              {               
                  int temp=num[j]+num[k];
                  if(temp==-first)
                  {
                      vector<int> tempResult; 
                      tempResult.push_back(first);
                      tempResult.push_back(num[j]);
                      tempResult.push_back(num[k]);
                      result.push_back(tempResult); 
                      j++; 
                      k--; 
                  }else if(temp<-first)
                  {
                      j++; 
                  }else 
                  {
                      k--;
                  }
                    while(j-1!=i && num[j]==num[j-1])
                      {
                          j++; 
                      }
                    while(k+1!=num.size() && num[k]==num[k+1])
                      {
                          k--; 
                      }
              }
          }
          return result; 
    }
};

File: 3Sum.java
public class Solution {
    public ArrayList<ArrayList<Integer>> threeSum(int[] num) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ArrayList<ArrayList<Integer>> result=new ArrayList<ArrayList<Integer>>(); 
        if(num.length<3) return result; 
        Arrays.sort(num);
        for(int i=0; i<=num.length-3; i++)
        {
            if(i!=0 && num[i]==num[i-1]) continue; 
            int sum=-num[i]; 
            int start=i+1, end=num.length-1;  
            while(start<end)
            {
                int temp=num[start]+num[end];
                if(temp==sum)
                {
                    ArrayList<Integer> list=new ArrayList<Integer>(); 
                    list.add(num[i]);
                    list.add(num[start]);
                    list.add(num[end]);
                    result.add(list);
                    start++; 
                    while(start<end && num[start]==num[start-1])
                    {
                        start++; 
                    }
                    end--; 
                    while(start<end && num[end]==num[end+1])
                    {
                        end--; 
                    }
                }else if(temp>sum)
                {
                    end--; 
                }else
                {
                    start++; 
                }
            }
        }
        return result; 
    }
    
}
